cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,380 @@
+import com.linkedin.kafka.cruisecontrol.analyzer.BalancingConstraint;
+import com.linkedin.kafka.cruisecontrol.analyzer.OptimizationOptions;
+import com.linkedin.kafka.cruisecontrol.analyzer.ProvisionRecommendation;
+import com.linkedin.kafka.cruisecontrol.analyzer.ProvisionResponse;
+import com.linkedin.kafka.cruisecontrol.analyzer.ProvisionStatus;
+import com.linkedin.kafka.cruisecontrol.config.BrokerSetDataHelper;
+import com.linkedin.kafka.cruisecontrol.config.BrokerSetDataStore;
+import com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException;
+import com.linkedin.kafka.cruisecontrol.model.Broker;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModel;
+import com.linkedin.kafka.cruisecontrol.model.Partition;
+import com.linkedin.kafka.cruisecontrol.model.Replica;
+import com.linkedin.kafka.cruisecontrol.model.ReplicaSortFunctionFactory;
+import com.linkedin.kafka.cruisecontrol.model.SortedReplicasHelper;
+import com.linkedin.kafka.cruisecontrol.monitor.ModelCompletenessRequirements;
+
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.ACCEPT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.BROKER_REJECT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.REPLICA_REJECT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.GoalUtils.MIN_NUM_VALID_WINDOWS_FOR_SELF_HEALING;
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.GoalUtils.replicaSortName;
+
+
+/**
+ * HARD GOAL: Generate proposals to provide brokerSet-aware replica distribution.
+ *
+ * <p>
+ * BrokerSets are defined as a group of Brokers within a cluster. With BrokerSets, a topic can be placed within the boundary of a BrokerSet.
+ * BrokerSets help reduce blast radius and provide a layer of isolation within a Kafka Cluster. BrokerSets can expand their boundaries or shrink.
+ * </p>
+ * <p>
+ * Assuming a case where RF is 3, the distribution of topic-partitions can look like this :
+ *
+ *
+ * <pre>
+ * BrokerSet RED                | BrokerSet BLUE              | BrokerSet GREEN
+ * -------------------------------------------------------------------------------
+ * Broker1  t1p1-0,  t2p1-2     | Broker4 t3p1-1              | Broker7
+ * Broker2  t2p1-0,  t1p1-1     | Broker5 t3p1-2              | Broker8
+ * Broker3  t1p1-2,  t2p1-1     | Broker6 t3p1-0              | Broker9
+ *
+ * </pre>
+ * </p>
+ *
+ * <p>
+ * There could also be some topics that do not follow the BrokerSet boundary.
+ * Such topics should be excluded from movements using the {topics.excluded.from.partition.movement} property
+ *
+ * Supported Optimization Options are :
+ *
+ * _excludedTopics -> Handled by BrokerSetAwareGoal
+ * _excludedBrokersForReplicaMove -> Handled by AbstractGoal
+ * _requestedDestinationBrokerIds -> Handled by AbstractGoal
+ * _onlyMoveImmigrantReplicas -> Handled by BrokerSetAwareGoa
+ * </p>
+ *
+ * <p>
+ * The algorithm to Optimize Goals is :
```

Comment:
This describes the optimizer algorithm - not specific to BrokerSetAwareGoal. I have added it so that the reader of BrokerSetAwareGoal would understand the whole picture.